### ISSUE TITLE
Make exceptions more specific

### ISSUE THREAD

**Issue by valgur** [sentinelsat/sentinelsat#131]
I first suggested this change in #108.

I would split `SentinelAPIError` into a couple of more specific ones:

  - `SentinelAPIError`: for relaying API responses that return an exception, e.g. for invalid query parameters. The response is expected to contain the `cause-message` in that case (as in #107).
  - `UnauthorizedError`: for cases where invalid credentials were provided or when the server does not accept them yet for whatever reason (a common issue for new users, I think).
  - `ServerError`: for cases when the server is unreachable, under maintenance or the response is just unparsable garbage.

This would enable different error cases to be caught more elegantly. E.g. you probably just want to retry your query after some delay in case of `ServerError`.

**Comment by gkeller2** [comment]
Started working on this today. I'll have an early version of the change soon to discuss.

**Comment by Andrey-Raspopov** [comment]
Is this active? If it is, I can try to divide them.

**Comment by Fernerkundung** [comment]
@Andrey-Raspopov I think we can reasonably assume that @gkeller2 isn't working on it anymore a year later. Go for it 😃

**Comment by Andrey-Raspopov** [comment]
Are there any default project credentials for scihub? I've registered for API access a week ago and still can't run a query.

**Comment by Fernerkundung** [comment]
There are no default project credentials, as account sharing is forbidden by the SciHub terms and conditions.

The propagation of accounts to the apihub used to take a while. If you test with `https://scihub.copernicus.eu/dhus/` the registered accounts work right away (just trested with a new one).

**Comment by Andrey-Raspopov** [comment]
Yeah, it works there, and it worked since day one, so I suppose I'll just wait more:)

**Comment by Fernerkundung** [comment]
No need to wait - the API for both endpoints is identical. You can test with '/dhus' instead of `/apihub` locally without any issues.

**Comment by dwlsalmeida** [comment]
Is someone currently working on this or is it still an issue?

I'd like to work on that, seems like a good first issue.

**Comment by valgur** [comment]
It's definitely up for grabs, go ahead!

Creating realistic mock responses for unit tests might be somewhat tricky, especially for `ServerError`, but the existing tests should provide some useful examples. You will likely need to use something like [`requests-mock`](https://requests-mock.readthedocs.io/en/latest/index.html) to set up test cases for `ServerError`.

**Comment by dwlsalmeida** [comment]
I have some of the code ready, but so far I've been unable to run the tests.

Here's the stacktrace for `py.test -v -vcr disable`

```Traceback (most recent call last):
  File "/usr/bin/py.test", line 11, in <module>
    load_entry_point('pytest==4.4.0', 'console_scripts', 'py.test')()
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 60, in main
    config = _prepareconfig(args, plugins)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 201, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/manager.py", line 68, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/manager.py", line 62, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/helpconfig.py", line 93, in pytest_cmdline_parse
    config = outcome.get_result()
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/dwls/.local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 679, in pytest_cmdline_parse
    self.parse(args)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 896, in parse
    self._preparse(args, addopts=addopts)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 829, in _preparse
    self._initini(args)
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 752, in _initini
    config=self,
  File "/home/dwls/.local/lib/python3.7/site-packages/_pytest/config/findpaths.py", line 103, in determine_setup
    iniconfig = py.iniconfig.IniConfig(inifile)
  File "/home/dwls/.local/lib/python3.7/site-packages/py/_vendored_packages/iniconfig.py", line 52, in __init__
    f = open(self.path)
FileNotFoundError: [Errno 2] No such file or directory: 'r'
```

By the way, "pip install py" fails with a timeout everytime, and although I have manually installed python-py from my linux distribution, a simple

`import py` gives out an import error


Would you please help?

**Comment by valgur** [comment]
Try using `pytest` instead of `py.test` (https://stackoverflow.com/a/41893170/2997179) and use two hyphens for the `--vcr` option. I suspect the error is caused by the latter typo.

**Comment by valgur** [comment]
Implemented in #285.
